### ISSUE TITLE
A race condition issue in `postcss-custom-properties`

### ISSUE THREAD

**Issue by malash** [csstools/postcss-plugins#331]
### Bug description

Recently I found a race condition issue of `postcss-custom-properties` while using it with other async PostCSS plugins.

I cost a lot of time to figure out the root issue and finally I can provide a easy way to reproduce this bug.

Full example repo: https://github.com/malash/postcss-plugins-issue-331

To reproduce this issue you need:

1. Use `postcss-custom-properties` with another async plugin ( which means the listeners of plugin should return a promise ).

I implemented this fake plugin:

```tsx
// this fake async plugin do nothing but await for next tick
const fakeAsyncPlugin = (opts = {}) => {
  // Plugin creator to check options or prepare caches
  return {
    postcssPlugin: 'A fake async plugin',
    Declaration: async () => {
      await new Promise(resolve => setTimeout(resolve, 0));
    },
  };
};
fakeAsyncPlugin.postcss = true;
```

2. Create a PostCSS instance and make sure the async runs before `postcss-custom-properties`.

```tsx
const postcssInstance = postcss([
  fakeAsyncPlugin(),
  postcssCustomProperties({
    preserve: false,
  }),
]);
```

3. Call `process` method in parallel. Make sure one CSS contains custom properties and others doesn't.

```tsx
const WITH_CUSTOM_PROPERTIES = `
:root {
  --color: red;
}

h1 {
  color: var(--color);
}
`;

const WITHOUT_CUSTOM_PROPERTIES = `
div {
  height: 10px;
}
`;

await Promise.all([
  postcssInstance.process(WITH_CUSTOM_PROPERTIES);
  postcssInstance.process(WITHOUT_CUSTOM_PROPERTIES);
  postcssInstance.process(WITHOUT_CUSTOM_PROPERTIES);
  postcssInstance.process(WITHOUT_CUSTOM_PROPERTIES);
  ...
]);
```

4. Check the output of `WITH_CUSTOM_PROPERTIES`, the `var(--color)` is not replaced as expected.

> In my case I use PostCSS with Webpack, the `postcss-loader` usually run in parallel, that's why I fall into this bug.

### Root issue and solution

I believe the root cause is this line:

https://github.com/csstools/postcss-plugins/blob/e9893a0d0d1fc8ed59efa843c5eff2750cc081f3/plugins/postcss-custom-properties/src/index.ts#L51

The plugin function is initialed only once in the `postcss([pluginFn()])` call. But the `prepare` listener runs for each CSS input file. In this plugin, the `customProperties` should be placed `prepare function.

```tsx
// let customProperties: Map<string, valuesParser.ParsedValue> = new Map();

return {
  postcssPlugin: 'postcss-custom-properties',
  prepare() {
    // move to here
    let customProperties: Map<string, valuesParser.ParsedValue> = new Map();
    return {
      Once:() => {},
      Declaration: () => {},
    };
  },
};
```

### Source CSS

```css
:root {
  --color: red;
}

h1 {
  color: var(--color);
}
```

### Expected CSS

```css
h1 {
  color: red;
}
```

### Actual CSS

```css
:root {
  --color: red;
}

h1 {
  color: var(--color);
}
```

### Does it happen with `npx @csstools/csstools-cli <plugin-name> minimal-example.css`?

No

### Plugin version

12.1.6

### What OS are you experiencing this on?

macOS

### Node Version

16.13.0

### Validations

- [X] Follow our [Code of Conduct](https://github.com/csstools/postcss-plugins/blob/master/CODE_OF_CONDUCT.md)
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Would you like to open a PR for this bug?

- [X] I'm willing to open a PR

**Comment by romainmenke** [comment]
Thank you for reporting this!

I don't think you even need the second async plugin to trigger this.
The plugin isn't safe to use in this way :/

At the moment it only supports reuse of the same instance in series, not in parallel.
We will have to take a deep dive to see if this can be improved.

To use the plugin in parallel you can create multiple instances.

**Comment by romainmenke** [comment]
My current gut feeling is that everything should be moved into `prepare`.
But I am unsure of what the side effects of this will be.

**Comment by malash** [comment]
@romainmenke https://github.com/csstools/postcss-plugins/pull/332 😄

**Comment by Antonio-Laguna** [comment]
@malash this has been released as `12.1.7`. Thanks a lot for doing this!
